code/data-preprocessing.py

```diff
@@ -66,40 +37,138 @@ def checkDigit(s):
-     checkDigit("03024350799") 
+    SAMPLE USAGE
+    ------------
+    # check digit is 8
+    checkDigit("03024350799")
     """
     even_sum = 0
     odd_sum = 0
     for i in range(len(s)):
-        if i%2 == 0:
+        if i % 2 == 0:
             odd_sum += int(s[i])
         else: 
             even_sum += int(s[i])
-    remainder = (odd_sum*3 + even_sum)%10
+    remainder = (odd_sum * 3 + even_sum) % 10
     if remainder == 0:
         return 0
     else:
         return 10 - remainder
 
+    
 def UPC10to12(s):
     """
-     Returns list of possible 12-digit UPCs for a 10-digit UPC given each possible 'Number system digit'
-     (No more than one UPC in the list actually exists)
+    Returns list of possible 12-digit UPCs for a 10-digit UPC given each
+    possible 'Number system digit' 
 
+    (No more than one UPC in the list actually exists)
 
-     SAMPLE USAGE
-     ------------
-     #first element of the returned list '030243507998' is the actual UPC
-     UPC10to12("3024350799")
-        
+    SAMPLE USAGE
+    ------------
+    #first element of the returned list '030243507998' is the actual UPC
+    UPC10to12("3024350799")
     """
     upc12_possible_list = []
-    for num in [0,1,6,7,8]:
+    for num in [0, 1, 6, 7, 8]:
         upc_11 = str(num)+s
         upc_12 = upc_11 + str(checkDigit(upc_11))
         upc12_possible_list.append(upc_12)
-    return upc12_possible_list;
+    return upc12_possible_list
+
+
+def getASIN(upc):
+    """
+    Query UPCtoASIN.com to determine ASIN from UPC.
+
+    If the UPC is not twelve digits, the function will throw a ValueError.
+    
+    If the UPC is twelve digits, query the website and retrieve ASIN, then wait
+    one second (per API rate limit).
+
+    SAMPLE USAGE
+    ------------
+    getASIN("876063002233")     # 'B001BCH7KM'
+    getASIN("8760630022")       # Throws an error
+    """
+    if len(upc) != 12:
+        raise ValueError("UPC must be twelve digits long")
+    else:
+        url = "http://upctoasin.com/" + upc
+        response = requests.get(url)
+        sleep(1)                # Sleep for one second
+        return(response.text)
+
+    
+def searchPossUPCs(upc_list):
+    """
+    Look for an Amazon ID from a list of possible UPCs.
+
+    For a list of candidate UPCs generated by UPC10to12, loop through and
+    search for an Amazon ID (ASIN). If an ASIN is found, the loop ends and the
+    ASIN is returned, otherwise the function returns "UPCNOTFOUND".
+
+    This function throws an error if it hits a UPC whose length is not 12.
+
+    SAMPLE USAGE
+    ------------
+    a = ['076063002237', '176063002234', '676063002239', '776063002236',
+         '876063002233']
+    searchPossUPCs(a)
+
+    # Throws an error because b[2] is not 12 digits
+    b = ['076063002237', '176063002234', '124542', '676063002239',
+         '776063002236', '876063002233']
+    searchPossUPCs(a)
+
+    """
+    res = None
+    for code in upc_list:
+        if len(code) != 12:
```

Comment:
@mikemunsell pointed out that this is redundant, since `getASIN` also checks UPC length. Not sure if it's really a problem to have it here or not but I'm open to suggestions.